{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import gc\n",
    "import os\n",
    "import time\n",
    "\n",
    "# Left side barcodes\n",
    "lbc = {\"CCTAAACTACGG\":\"F-01\",\n",
    "\"TGCAGATCCAAC\":\"F-02\",\n",
    "\"CCATCACATAGG\":\"F-03\",\n",
    "\"GTGGTATGGGAG\":\"F-04\",\n",
    "\"ACTTTAAGGGTG\":\"F-05\",\n",
    "\"GAGCAACATCCT\":\"F-06\",\n",
    "\"TGTTGCGTTTCT\":\"F-07\",\n",
    "\"ATGTCCGACCAA\":\"F-08\",\n",
    "\"AGGTACGCAATT\":\"F-09\",\n",
    "\"ACAGCCACCCAT\":\"F-10\",\n",
    "\"TGTCTCGCAAGC\":\"F-11\",\n",
    "\"GAGGAGTAAAGC\":\"F-12\",\n",
    "\"GTTACGTGGTTG\":\"F-13\",\n",
    "\"TACCGCCTCGGA\":\"F-14\",\n",
    "\"CGTAAGATGCCT\":\"F-15\",\n",
    "\"TACCGGCTTGCA\":\"F-16\",\n",
    "\"ATCTAGTGGCAA\":\"F-17\",\n",
    "\"CCAGGGACTTCT\":\"F-18\",\n",
    "\"CACCTTACCTTA\":\"F-19\",\n",
    "\"ATAGTTAGGGCT\":\"F-20\",\n",
    "\"GCACTTCATTTC\":\"F-21\",\n",
    "\"TTAACTGGAAGC\":\"F-22\",\n",
    "\"CGCGGTTACTAA\":\"F-23\",\n",
    "\"GAGACTATATGC\":\"F-24\",}\n",
    "\n",
    "rbc = {\"CCGTAGTTTAGG\":\"R-01\",\n",
    "\"GTTGGATCTGCA\":\"R-02\",\n",
    "\"CCTATGTGATGG\":\"R-03\",\n",
    "\"CTCCCATACCAC\":\"R-04\",\n",
    "\"CACCCTTAAAGT\":\"R-05\",\n",
    "\"AGGATGTTGCTC\":\"R-06\",\n",
    "\"AGAAACGCAACA\":\"R-07\",\n",
    "\"TTGGTCGGACAT\":\"R-08\",\n",
    "\"AATTGCGTACCT\":\"R-09\",\n",
    "\"ATGGGTGGCTGT\":\"R-10\",\n",
    "\"GCTTGCGAGACA\":\"R-11\",\n",
    "\"GCTTTACTCCTC\":\"R-12\",\n",
    "\"CAACCACGTAAC\":\"R-13\",\n",
    "\"TCCGAGGCGGTA\":\"R-14\",\n",
    "\"AGGCATCTTACG\":\"R-15\",\n",
    "\"TGCAAGCCGGTA\":\"R-16\",}\n",
    "\n",
    "\"\"\"\n",
    "# Right side barcodes\n",
    "rbc = {\"CCTAAACTACGG\":\"R-01\",\n",
    "\"TGCAGATCCAAC\":\"R-02\",\n",
    "\"CCATCACATAGG\":\"R-03\",\n",
    "\"GTGGTATGGGAG\":\"R-04\",\n",
    "\"ACTTTAAGGGTG\":\"R-05\",\n",
    "\"GAGCAACATCCT\":\"R-06\",\n",
    "\"TGTTGCGTTTCT\":\"R-07\",\n",
    "\"ATGTCCGACCAA\":\"R-08\",\n",
    "\"AGGTACGCAATT\":\"R-09\",\n",
    "\"ACAGCCACCCAT\":\"R-10\",\n",
    "\"TGTCTCGCAAGC\":\"R-11\",\n",
    "\"GAGGAGTAAAGC\":\"R-12\",\n",
    "\"GTTACGTGGTTG\":\"R-13\",\n",
    "\"TACCGCCTCGGA\":\"R-14\",\n",
    "\"CGTAAGATGCCT\":\"R-15\",\n",
    "\"TACCGGCTTGCA\":\"R-16\",}\"\"\"\n",
    "\n",
    "samples = {\n",
    "\"F-01R-01\":\"XA1_T0\",\n",
    "\"F-01R-02\":\"XA2_T0\",\n",
    "\"F-01R-03\":\"XA3_T0\",\n",
    "\"F-01R-04\":\"XA4_T0\",\n",
    "\"F-01R-05\":\"XB1_T0\",\n",
    "\"F-01R-06\":\"XB2_T0\",\n",
    "\"F-01R-07\":\"XB3_T0\",\n",
    "\"F-01R-08\":\"XB4_T0\",\n",
    "\"F-02R-01\":\"XC1_T0\",\n",
    "\"F-02R-02\":\"XC2_T0\",\n",
    "\"F-02R-03\":\"XC3_T0\",\n",
    "\"F-02R-04\":\"XC4_T0\",\n",
    "\"F-02R-05\":\"DA1_T0\",\n",
    "\"F-02R-06\":\"DA2_T0\",\n",
    "\"F-02R-07\":\"DA3_T0\",\n",
    "\"F-02R-08\":\"DA4_T0\",\n",
    "\"F-03R-01\":\"DB1_T0\",\n",
    "\"F-03R-02\":\"DB2_T0\",\n",
    "\"F-03R-03\":\"DB3_T0\",\n",
    "\"F-03R-04\":\"DB4_T0\",\n",
    "\"F-03R-05\":\"DC1_T0\",\n",
    "\"F-03R-06\":\"DC2_T0\",\n",
    "\"F-03R-07\":\"DC3_T0\",\n",
    "\"F-03R-08\":\"DC4_T0\",\n",
    "\"F-04R-01\":\"QA1_T0\",\n",
    "\"F-04R-02\":\"QA2_T0\",\n",
    "\"F-04R-03\":\"QA3_T0\",\n",
    "\"F-04R-04\":\"QA4_T0\",\n",
    "\"F-04R-05\":\"QB1_T0\",\n",
    "\"F-04R-06\":\"QB2_T0\",\n",
    "\"F-04R-07\":\"QB3_T0\",\n",
    "\"F-04R-08\":\"QB4_T0\",\n",
    "\"F-05R-01\":\"QC1_T0\",\n",
    "\"F-05R-02\":\"QC2_T0\",\n",
    "\"F-05R-03\":\"QC3_T0\",\n",
    "\"F-05R-04\":\"QC4_T0\",\n",
    "\"F-05R-05\":\"DQA1_T0\",\n",
    "\"F-05R-06\":\"DQA2_T0\",\n",
    "\"F-05R-07\":\"DQA3_T0\",\n",
    "\"F-05R-08\":\"DQA4_T0\",\n",
    "\"F-06R-01\":\"DQB1_T0\",\n",
    "\"F-06R-02\":\"DBQ2_T0\",\n",
    "\"F-06R-03\":\"DBQ3_T0\",\n",
    "\"F-06R-04\":\"DQB4_T0\",\n",
    "\"F-06R-05\":\"DQC1_T0\",\n",
    "\"F-06R-06\":\"DQC2_T0\",\n",
    "\"F-06R-07\":\"DQC3_T0\",\n",
    "\"F-06R-08\":\"DQC4_T0\",\n",
    "\"F-07R-01\":\"XA1_T1\",\n",
    "\"F-07R-02\":\"XA2_T1\",\n",
    "\"F-07R-03\":\"XA3_T1\",\n",
    "\"F-07R-04\":\"XA4_T1\",\n",
    "\"F-07R-05\":\"XB1_T1\",\n",
    "\"F-07R-06\":\"XB2_T1\",\n",
    "\"F-07R-07\":\"XB3_T1\",\n",
    "\"F-07R-08\":\"XB4_T1\",\n",
    "\"F-08R-01\":\"XC1_T1\",\n",
    "\"F-08R-02\":\"XC2_T1\",\n",
    "\"F-08R-03\":\"XC3_T1\",\n",
    "\"F-08R-04\":\"XC4_T1\",\n",
    "\"F-08R-05\":\"DA1_T1\",\n",
    "\"F-08R-06\":\"DA2_T1\",\n",
    "\"F-08R-07\":\"DA3_T1\",\n",
    "\"F-08R-08\":\"DA4_T1\",\n",
    "\"F-09R-01\":\"DB1_T1\",\n",
    "\"F-09R-02\":\"DB2_T1\",\n",
    "\"F-09R-03\":\"DB3_T1\",\n",
    "\"F-09R-04\":\"DB4_T1\",\n",
    "\"F-09R-05\":\"DC1_T1\",\n",
    "\"F-09R-06\":\"DC2_T1\",\n",
    "\"F-09R-07\":\"DC3_T1\",\n",
    "\"F-09R-08\":\"DC4_T1\",\n",
    "\"F-10R-01\":\"QA1_T1\",\n",
    "\"F-10R-02\":\"QA2_T1\",\n",
    "\"F-10R-03\":\"QA3_T1\",\n",
    "\"F-10R-04\":\"QA4_T1\",\n",
    "\"F-10R-05\":\"QB1_T1\",\n",
    "\"F-10R-06\":\"QB2_T1\",\n",
    "\"F-10R-07\":\"QB3_T1\",\n",
    "\"F-10R-08\":\"QB4_T1\",\n",
    "\"F-11R-01\":\"QC1_T1\",\n",
    "\"F-11R-02\":\"QC2_T1\",\n",
    "\"F-11R-03\":\"QC3_T1\",\n",
    "\"F-11R-04\":\"QC4_T1\",\n",
    "\"F-11R-05\":\"DQA1_T1\",\n",
    "\"F-11R-06\":\"DQA2_T1\",\n",
    "\"F-11R-07\":\"DQA3_T1\",\n",
    "\"F-11R-08\":\"DQA4_T1\",\n",
    "\"F-12R-01\":\"DQB1_T1\",\n",
    "\"F-12R-02\":\"DBQ2_T1\",\n",
    "\"F-12R-03\":\"DBQ3_T1\",\n",
    "\"F-12R-04\":\"DQB4_T1\",\n",
    "\"F-12R-05\":\"DQC1_T1\",\n",
    "\"F-12R-06\":\"DQC2_T1\",\n",
    "\"F-12R-07\":\"DQC3_T1\",\n",
    "\"F-12R-08\":\"DQC4_T1\",\n",
    "\"F-13R-01\":\"XA1_T2\",\n",
    "\"F-13R-02\":\"XA2_T2\",\n",
    "\"F-13R-03\":\"XA3_T2\",\n",
    "\"F-13R-04\":\"XA4_T2\",\n",
    "\"F-13R-05\":\"XB1_T2\",\n",
    "\"F-13R-06\":\"XB2_T2\",\n",
    "\"F-13R-07\":\"XB3_T2\",\n",
    "\"F-13R-08\":\"XB4_T2\",\n",
    "\"F-14R-01\":\"XC1_T2\",\n",
    "\"F-14R-02\":\"XC2_T2\",\n",
    "\"F-14R-03\":\"XC3_T2\",\n",
    "\"F-14R-04\":\"XC4_T2\",\n",
    "\"F-14R-05\":\"DA1_T2\",\n",
    "\"F-14R-06\":\"DA2_T2\",\n",
    "\"F-14R-07\":\"DA3_T2\",\n",
    "\"F-14R-08\":\"DA4_T2\",\n",
    "\"F-15R-01\":\"DB1_T2\",\n",
    "\"F-15R-02\":\"DB2_T2\",\n",
    "\"F-15R-03\":\"DB3_T2\",\n",
    "\"F-15R-04\":\"DB4_T2\",\n",
    "\"F-15R-05\":\"DC1_T2\",\n",
    "\"F-15R-06\":\"DC2_T2\",\n",
    "\"F-15R-07\":\"DC3_T2\",\n",
    "\"F-15R-08\":\"DC4_T2\",\n",
    "\"F-16R-01\":\"QA1_T2\",\n",
    "\"F-16R-02\":\"QA2_T2\",\n",
    "\"F-16R-03\":\"QA3_T2\",\n",
    "\"F-16R-04\":\"QA4_T2\",\n",
    "\"F-16R-05\":\"QB1_T2\",\n",
    "\"F-16R-06\":\"QB2_T2\",\n",
    "\"F-16R-07\":\"QB3_T2\",\n",
    "\"F-16R-08\":\"QB4_T2\",\n",
    "\"F-17R-01\":\"QC1_T2\",\n",
    "\"F-17R-02\":\"QC2_T2\",\n",
    "\"F-17R-03\":\"QC3_T2\",\n",
    "\"F-17R-04\":\"QC4_T2\",\n",
    "\"F-17R-05\":\"DQA1_T2\",\n",
    "\"F-17R-06\":\"DQA2_T2\",\n",
    "\"F-17R-07\":\"DQA3_T2\",\n",
    "\"F-17R-08\":\"DQA4_T2\",\n",
    "\"F-18R-01\":\"DQB1_T2\",\n",
    "\"F-18R-02\":\"DBQ2_T2\",\n",
    "\"F-18R-03\":\"DBQ3_T2\",\n",
    "\"F-18R-04\":\"DQB4_T2\",\n",
    "\"F-18R-05\":\"DQC1_T2\",\n",
    "\"F-18R-06\":\"DQC2_T2\",\n",
    "\"F-18R-07\":\"DQC3_T2\",\n",
    "\"F-18R-08\":\"DQC4_T2\",\n",
    "\"F-19R-01\":\"BacHo_DA\",\n",
    "\"F-19R-02\":\"BacHo_DB\",\n",
    "\"F-19R-03\":\"BacHo_DC\",\n",
    "\"F-19R-04\":\"BacHo_DQA\",\n",
    "\"F-19R-05\":\"BacHo_DQB\",\n",
    "\"F-19R-06\":\"BacHo_DQC\",\n",
    "\"F-19R-07\":\"H2_Metatrans14\",\n",
    "\"F-19R-08\":\"H3_Metatrans14\",\n",
    "\"F-20R-01\":\"H4_Metatrans14\",\n",
    "\"F-20R-02\":\"H5_Metatrans14\",\n",
    "\"F-20R-03\":\"H6_Metatrans14\",\n",
    "\"F-20R-04\":\"H7_Metatrans14\",\n",
    "\"F-20R-05\":\"H8_Metatrans14\",\n",
    "\"F-20R-06\":\"D2_Metatrans14\",\n",
    "\"F-20R-07\":\"D3_Metatrans14\",\n",
    "\"F-20R-08\":\"D4_Metatrans14\",\n",
    "\"F-21R-01\":\"D5_Metatrans14\",\n",
    "\"F-21R-02\":\"D6_Metatrans14\",\n",
    "\"F-21R-03\":\"D7_Metatrans14\",\n",
    "\"F-21R-04\":\"D8_Metatrans14\",\n",
    "\"F-21R-05\":\"PE1_doseH\",\n",
    "\"F-21R-06\":\"PE1_doseD\",\n",
    "\"F-21R-07\":\"PE2_doseH\",\n",
    "\"F-21R-08\":\"PE2_doseD\",\n",
    "\"F-22R-01\":\"619_D_Homo\",\n",
    "\"F-22R-02\":\"619_H_Homo\",\n",
    "\"F-22R-03\":\"R_CK4_G4_F1\",\n",
    "\"F-22R-04\":\"R_CK4_G6_F2\",\n",
    "\"F-22R-05\":\"R_POPA_G2_F1\",\n",
    "\"F-22R-06\":\"R_POPA_G3_F2\",\n",
    "\"F-22R-07\":\"S_CK6_G1_F1\",\n",
    "\"F-22R-08\":\"S_POPA_G6_F2\",\n",
    "\"F-23R-01\":\"S_POPA_G7_F2\",\n",
    "\"F-23R-02\":\"623_D_Homo\",\n",
    "\"F-23R-03\":\"623_H_Homo\",\n",
    "\"F-23R-04\":\"R_CK4_G4_F2\",\n",
    "\"F-23R-05\":\"R_CK4_G6_F3\",\n",
    "\"F-23R-06\":\"R_POPA_G2_F2\",\n",
    "\"F-23R-07\":\"R_POPA_G3_F3\",\n",
    "\"F-23R-08\":\"S_CK6_G1_F2\",\n",
    "\"F-24R-01\":\"S_CK6_G7_F4\",\n",
    "\"F-24R-02\":\"S_POPA_G6_F3\",\n",
    "\"F-24R-03\":\"S_POPA_G7_F5\",\n",
    "\"F-24R-04\":\"R_CK4_G4_F3\",\n",
    "\"F-24R-05\":\"R_CK4_G6_F5\",\n",
    "\"F-24R-06\":\"R_POPA_G3_F4\",\n",
    "\"F-24R-07\":\"S_CK6_G1_F3\",\n",
    "\"F-24R-08\":\"S_CK6_G7_F5\",}\n",
    "\n",
    "# Can also be passed as arguments\n",
    "fadaptor = \"ACGCTCTTCCGATCT\"\n",
    "radaptor = \"AGATCGGAAGAGC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/eho/Documents/becca\n",
      "Combine Barcodes.ipynb R2_test.fastq          left.fastq             test.fastq\n",
      "\u001b[34mMiSeq16S\u001b[m\u001b[m               demultiplex.py         right.fastq            \u001b[34mtest_out\u001b[m\u001b[m\n",
      "R1_RevComp_test.fastq  demultiplex.py~        short\n"
     ]
    }
   ],
   "source": [
    "!pwd\n",
    "!ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def process(left, right, outdir, filter_limit, filter_length):\n",
    "    \"\"\"\n",
    "    Process the fastq file. For each record (read), we do the following:\n",
    "    \n",
    "    1) Trim the adaptors\n",
    "    2) Remove the barcodes\n",
    "    3) Lookup the appropriate file, filter if necessary\n",
    "    4) Queue the file for writing\n",
    "    \"\"\"\n",
    "    \n",
    "    print \"reading from: \" \n",
    "    print \"left: \" + left\n",
    "    print \"right: \" + right\n",
    "    print \"writing to: \" + outdir\n",
    "    print \"limit: \" + str(filter_limit)\n",
    "    \n",
    "    \n",
    "    # Create a set of file names, a list of lines to be written, and line counts\n",
    "    left_files = {}\n",
    "    right_files = {}\n",
    "\n",
    "    prepare_files(\"left\", o, left_files)\n",
    "    prepare_files(\"right\", o, right_files)\n",
    "    \n",
    "    count = 0\n",
    "    left_filtered = 0\n",
    "    right_filtered = 0\n",
    "    ftrimmed = 0\n",
    "    rtrimmed = 0\n",
    "    \n",
    "    right_primer_dimers = 0\n",
    "    left_primer_dimers = 0\n",
    "    \n",
    "    left_data = SeqIO.parse(left,\"fastq\")\n",
    "    right_data = SeqIO.parse(right,\"fastq\")\n",
    "    \n",
    "    start = time.time()\n",
    "    \n",
    "    left_read = left_data.next();\n",
    "    right_read = right_data.next();\n",
    "    while (left_read or right_read):\n",
    "        \n",
    "        if (not left_read) or (not right_read):\n",
    "            raise Exception(\"This isn't right? More reads in one that the other\")\n",
    "       \n",
    "        count += 1\n",
    "        if count % 25000 == 0 and count > 1:\n",
    "            print \"processed {0} reads in {1} seconds ({2}% have been filtered from left, {3}% have been filtered from right)\".format(\\\n",
    "                str(count), str(time() - start), str(((left_filtered+0.0)/count)*100), str(((right_filtered+0.0)/count)*100))\n",
    "            print \"trimmed {0} front {1} rear\".format(ftrimmed, rtrimmed)\n",
    "        \n",
    "        # Trim the adaptor\n",
    "        left_read, fcheck, rt = trim_adaptors(left_read)\n",
    "        left_primer_dimers += rt\n",
    "            \n",
    "        right_read, ft, rcheck = trim_adaptors(right_read)\n",
    "        right_primer_dimers += ft\n",
    "            \n",
    "        ftrimmed += fcheck\n",
    "        rtrimmed += rcheck\n",
    "        \n",
    "        # Filter and find file\n",
    "        sample_name = get_sample_name(left_read, right_read)\n",
    "        left_sample_name = sample_name\n",
    "        right_sample_name = sample_name\n",
    "        \n",
    "        # Remove barcodes\n",
    "        left_read = left_read[12:]\n",
    "        right_read = right_read[:-12]\n",
    "    \n",
    "        # Also remove the primers\n",
    "        left_read = left_read[17:]\n",
    "        right_read = right_read[:-20]\n",
    "        \n",
    "        # Trim the reads\n",
    "        left_read = trim_left_read(left_read, filter_limit)\n",
    "        right_read = trim_right_read(right_read, filter_limit)\n",
    "        \n",
    "        # If the reads are too short throw them away\n",
    "        if (len(left_read) < filter_length):\n",
    "            left_sample_name = \"LOW_QUALITY_READS\"\n",
    "            left_filtered += 1\n",
    "        if (len(right_read) < filter_length):\n",
    "            right_sample_name = \"LOW_QUALITY_READS\"\n",
    "            right_filtered += 1\n",
    "\n",
    "        # format and append\n",
    "        format_lines(left_sample_name, right_sample_name, left_read, right_read, left_files, right_files)\n",
    "        \n",
    "        # Arbitrary... but seems to prevent memory overloads.\n",
    "        if count % 500000 == 0:\n",
    "            print \"DUMPING FILES... \"\n",
    "            dump_files(left_files)\n",
    "            dump_files(right_files)\n",
    "            \n",
    "        try:\n",
    "            left_read = left_data.next()\n",
    "            right_read = right_data.next()\n",
    "        except:\n",
    "            break\n",
    "            \n",
    "    print \"analysis complete, successfully processed {0} reads... writing to files\".format(str(count))\n",
    "    dump_files(left_files)\n",
    "    dump_files(right_files)\n",
    "        \n",
    "    print \"trimmed {0} front {1} rear\".format(ftrimmed, rtrimmed)\n",
    "    print \"primer dimers: {0} left, {1} right\".format(left_primer_dimers, right_primer_dimers)\n",
    "    \n",
    "    print \"files written, {0} reads processed, left filtered {1}, right filtered {2}\".format(\\\n",
    "                str(count),str(left_filtered),str(right_filtered))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def prepare_files(direction, outputfolder, file_map):\n",
    "    \"\"\"\n",
    "    Creates a set of files for Demultiplexing.\n",
    "    \"\"\"\n",
    "    loc = outputfolder + direction + \"/\"\n",
    "    \n",
    "    # create the folder directory if it doesn't exist\n",
    "    try:\n",
    "        os.stat(outputfolder)\n",
    "    except:\n",
    "        os.mkdir(outputfolder)\n",
    "        \n",
    "    try:\n",
    "        os.stat(loc)\n",
    "    except:\n",
    "        os.mkdir(loc)\n",
    "        \n",
    "    for key in samples:\n",
    "        name = samples[key]\n",
    "        t = loc + name + \".fasta\"\n",
    "        \n",
    "        # File name, lines list, lines written\n",
    "        file_map[name] = [t,[], 1]\n",
    "        \n",
    "    file_map[\"NO_MATCH\"] = [loc + \"NO_MATCH.fasta\", [], 1]\n",
    "    file_map[\"LOW_QUALITY_READS\"] = [loc + \"LOW_QUALITY_READS.fasta\", [], 1]\n",
    "    \n",
    "    for key in file_map:\n",
    "        f = open(file_map[key][0], 'w')\n",
    "        f.close()\n",
    "\n",
    "\n",
    "def dump_files(file_map):\n",
    "    \"\"\"\n",
    "    Dumps queued content into the appropriate file and garbage collects.\n",
    "    \"\"\"\n",
    "    for key in file_map:\n",
    "        t = file_map[key]\n",
    "        f = open(t[0],'a')\n",
    "        \n",
    "        for line in t[1]:\n",
    "            f.write(line)\n",
    "\n",
    "        file_map[key][1] = []\n",
    "        f.close()\n",
    "    \n",
    "    gc.collect()\n",
    "    \n",
    "\n",
    "def get_trim_index(rec):\n",
    "    \"\"\"\n",
    "    Computes the phred score of the record \n",
    "    \"\"\"\n",
    "    phreds = rec.letter_annotations[\"phred_quality\"]\n",
    "    \n",
    "    # Maximum possible score\n",
    "    low_score = 60\n",
    "    \n",
    "    for i in xrange(len(phreds)):\n",
    "        low_score = min(phreds[i], low_score)\n",
    "\n",
    "    return low_score\n",
    "\n",
    "\n",
    "def get_sample_name(leftread, rightread):\n",
    "    \"\"\"\n",
    "    Returns the filename in which to dump the record\n",
    "    \"\"\"\n",
    "    leftBC = leftread.seq[0:12]\n",
    "    rightBC = rightread.seq[-12:]\n",
    "    \n",
    "    if not ((leftBC in lbc) and (rightBC in rbc)):\n",
    "        return \"NO_MATCH\"\n",
    "        \n",
    "    ids = lbc[leftBC] + rbc[rightBC]\n",
    "    \n",
    "    if not ids in samples:\n",
    "        return \"NO_MATCH\"\n",
    "        \n",
    "    return samples[ids]\n",
    "\n",
    "\n",
    "def trim_adaptors(record):\n",
    "    \"\"\"\n",
    "    Removes the adaptors from the front and end of the read, if they exist.\n",
    "    \"\"\"\n",
    "    indexf = record.seq.find(fadaptor)\n",
    "    indexr = record.seq.find(radaptor)\n",
    "    \n",
    "    f = 0\n",
    "    r = 0\n",
    "    \n",
    "    # Trim the rear adaptor first if it's there, obvi\n",
    "    if indexr >= 0:\n",
    "        record = record[:indexr]\n",
    "        r = 1\n",
    "    if indexf >= 0: \n",
    "        len_fadaptor = len(fadaptor)\n",
    "        record = record[indexf+len_fadaptor:]\n",
    "        f = 1\n",
    "\n",
    "    return [record, f, r]\n",
    "\n",
    "\n",
    "def format_lines(left_sample_name, right_sample_name, left_read, right_read, left_files, right_files):\n",
    "    \"\"\"\n",
    "    Converts a fastq line into a fasta line, removes the adapters, and appends the header name to\n",
    "    the line.\n",
    "    \"\"\"\n",
    "    # Link to files\n",
    "    left_tupl = left_files[left_sample_name]\n",
    "    right_tupl = right_files[right_sample_name]\n",
    "    \n",
    "    left_line = format(left_read, \"fasta\")\n",
    "    left_line = \"{0}{1}_{2}:{3}\".format(left_line[0], left_sample_name, left_tupl[2], left_line[1:])\n",
    "    left_tupl[1].append(left_line)\n",
    "    left_tupl[2] += 1\n",
    "    \n",
    "    right_line = format(right_read, \"fasta\")\n",
    "    right_line = \"{0}{1}_{2}:{3}\".format(right_line[0], right_sample_name, right_tupl[2], right_line[1:])\n",
    "    right_tupl[1].append(right_line)\n",
    "    right_tupl[2] += 1\n",
    "    \n",
    "    \n",
    "def trim_left_read(left_read, filter_limit):  \n",
    "    \n",
    "    phreds = left_read.letter_annotations[\"phred_quality\"]\n",
    "    \n",
    "    for i in xrange(len(phreds)):\n",
    "        if (phreds[i] < filter_limit):\n",
    "            return left_read[:i]\n",
    "\n",
    "    return left_read\n",
    "\n",
    "\n",
    "def trim_right_read(right_read, filter_limit):\n",
    "    \n",
    "    phreds = right_read.letter_annotations[\"phred_quality\"]\n",
    "    \n",
    "    for i in xrange(len(phreds) - 1, 0, -1):\n",
    "        if (phreds[i] < filter_limit):\n",
    "            return right_read[i:]\n",
    "\n",
    "    return right_read\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading from: \n",
      "left: R2_test.fastq\n",
      "right: R1_RevComp_test.fastq\n",
      "writing to: test_out/\n",
      "limit: 20\n",
      "analysis complete, successfully processed 10000 reads... writing to files\n",
      "trimmed 0 front 238 rear\n",
      "primer dimers: 2114 left, 0 right\n",
      "files written, 10000 reads processed, left filtered 3354, right filtered 1503\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "//anaconda/envs/py27/lib/python2.7/site-packages/Bio/Seq.py:152: BiopythonWarning: Biopython Seq objects now use string comparison. Older versions of Biopython used object comparison. During this transition, please use hash(id(my_seq)) or my_dict[id(my_seq)] if you want the old behaviour, or use hash(str(my_seq)) or my_dict[str(my_seq)] for the new string hashing behaviour.\n",
      "  \"the new string hashing behaviour.\", BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "# This actually runs the script!\n",
    "\n",
    "# output directory\n",
    "o = \"test_out/\"\n",
    "\n",
    "# Open files for reading\n",
    "left = \"R2_test.fastq\"\n",
    "right = \"R1_RevComp_test.fastq\"\n",
    "#left = \"left.fastq\"\n",
    "#right = \"right.fastq\"\n",
    "\n",
    "process(left, right, o, 20, 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DC4_T2\n",
      "DC4_T0\n",
      "DQB4_T0\n",
      "DQC4_T2\n",
      "DQC2_T2\n"
     ]
    }
   ],
   "source": [
    "left = SeqIO.parse(\"R2_test.fastq\", \"fastq\")\n",
    "right = SeqIO.parse(\"R1_RevComp_test.fastq\", \"fastq\")\n",
    "\n",
    "left_out = open(\"beccas_left.fastq\", \"w\")\n",
    "right_out = open(\"beccas_right.fastq\", \"w\")\n",
    "\n",
    "left_seqs = []\n",
    "right_seqs = []\n",
    "for i in xrange(20):\n",
    "    left_r = left.next()\n",
    "    right_r = right.next()\n",
    "\n",
    "    left_read, fcheck, rt = trim_adaptors(left_r)\n",
    "    right_read, ft, rcheck = trim_adaptors(right_r)\n",
    "    \n",
    "    name = get_sample_name(left_read, right_read)\n",
    "    \n",
    "    if not name == \"NO_MATCH\":\n",
    "        print name\n",
    "        left_seqs.append(left_read)\n",
    "        right_seqs.append(right_read)\n",
    "        SeqIO.write(left_read, left_out, \"fastq\")\n",
    "        SeqIO.write(right_read, right_out, \"fastq\")\n",
    "        \n",
    "left_out.close()\n",
    "right_out.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[35, 31, 39, 39, 39, 36, 12, 35, 39, 34, 38, 38, 40, 39, 39, 39, 36, 34, 38, 40, 39, 39, 39, 40, 39, 39, 32, 31, 13, 32, 39, 38, 39, 39, 40, 39, 39, 40, 39, 40, 40, 40, 40, 40, 39, 39, 39, 35, 40, 40, 40, 40, 40, 39, 39, 39, 39, 39, 39, 38, 38, 40, 40, 39, 39, 39, 40, 38, 39, 39, 36, 38, 37, 36, 37, 39, 36, 37, 39, 40, 40, 40, 39, 38, 37, 39, 39, 39, 40, 40, 40, 39, 39, 39, 40, 39, 40, 39, 39, 40, 39, 39, 39, 37, 40, 39, 39, 39, 39, 39, 40, 39, 40, 40, 40, 40, 40, 40, 40, 40, 40, 39, 36, 39, 40, 40, 40, 39, 39, 39, 39, 36, 16, 40, 40, 40, 39, 40, 40, 40, 40, 39, 38, 40, 39, 40, 40, 39, 38, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 39, 40, 40, 40, 40, 40, 39, 38, 40, 39, 39, 40, 40, 39, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 39, 37, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 40, 35, 35, 35, 35, 35]\n"
     ]
    }
   ],
   "source": [
    "right_seqs[0].letter_annotations[\"phred_quality\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reading from: \n",
      "left: beccas_left.fastq\n",
      "right: beccas_right.fastq\n",
      "writing to: test_out/\n",
      "limit: 20\n",
      "analysis complete, successfully processed 5 reads... writing to files\n",
      "trimmed 0 front 0 rear\n",
      "primer dimers: 0 left, 0 right\n",
      "files written, 5 reads processed, left filtered 3, right filtered 3\n"
     ]
    }
   ],
   "source": [
    "# This actually runs the script!\n",
    "\n",
    "# output directory\n",
    "o = \"test_out/\"\n",
    "\n",
    "# Open files for reading\n",
    "left = \"beccas_left.fastq\"\n",
    "right = \"beccas_right.fastq\"\n",
    "#left = \"left.fastq\"\n",
    "#right = \"right.fastq\"\n",
    "\n",
    "process(left, right, o, 20, 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = \"!\\\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~\"\n",
    "\n",
    "a.index(\"E\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
